### PR TITLE
memoize seg numbers with getsegment2()

### PR DESCRIPTION
**Pull request by WalterBright**
Memoizing should make looking up a segment considerably faster.

### CODE REVIEW

File: src/dmd/backend/machobj.d [hunk 1]
```diff
@@ -248,6 +248,14 @@ int seg_tlsseg_bss = UNKNOWN;
...
  */
 int seg_tlsseg_data = UNKNOWN;
+
+int seg_cstring = UNKNOWN;        // __cstring section
+int seg_mod_init_func = UNKNOWN;  // __mod_init_func section
+int seg_mod_term_func = UNKNOWN;  // __mod_term_func section
+int seg_deh_eh = UNKNOWN;         // __deh_eh section
+int seg_textcoal_nt = UNKNOWN;
+int seg_tlscoal_nt = UNKNOWN;
+int seg_datacoal_nt = UNKNOWN;
```
Comment (andralex) [inline src/dmd/backend/machobj.d:258]: If any of these can't be negative, they ought to be `uint`. You only need one test for bounds checking etc.

File: src/dmd/backend/machobj.d [hunk 2]
```diff
@@ -1910,6 +1925,26 @@ int Obj_getsegment(const(char)* sectname, const(char)* segname,
...
+ *      flags = S_????
+ * Returns:
+ *      seg index
+ */
+private extern (D)
+int getsegment2(ref int seg, const(char)* sectname, const(char)* segname,
+        int align_, int flags)
+{
+    if (seg == UNKNOWN)
+        seg = Obj_getsegment(sectname, segname, align_, flags);
```
Comment (andralex) [inline src/dmd/backend/machobj.d:1944]: Wait, is `Obj_getsegment` only used from within this function? If so, can you make it static inside this one?

**Review by andralex — APPROVED** [review]

File: src/dmd/backend/machobj.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/backend/machobj.d:258]: UNKNOWN is -1. The OMF generation also uses negative seg numbers.

File: src/dmd/backend/machobj.d (hunk 2 above)
Comment (WalterBright) [reply src/dmd/backend/machobj.d:1944]: It's also used by dwarfdbginf.d

**Pull request merged**